### PR TITLE
[python-package] add plot tree

### PR DESCRIPTION
**Pull request by wxchan**
#171

### CODE REVIEW

File: python-package/lightgbm/plotting.py [hunk 1]
```diff
@@ -110,3 +121,94 @@ def plot_importance(booster, ax=None, height=0.2,
...
+
+def _to_graphviz(graph, tree_info):
+    """Convert specified tree to graphviz instance."""
+
+    def add(root, parent=None, decision=None):
+        """recursively add node or edge"""
+        if 'split_index' in root:  # non-leaf
+            name = 'split' + str(root['split_index'])
+            graph.node(name, label='threshold:' + str(root['split_feature']))
+            if root['decision_type'] == 'no_greater':
```
Comment (wxchan) [inline python-package/lightgbm/plotting.py]: Is the following logic right?
```
if 'decision_type' == 'no_greater':
  left_child < threshold
  right_child > threshold
else if 'decision_type' == 'is':
  left_child is threshold
  right_child isn't threshold
```

Comment (guolinke) [inline python-package/lightgbm/plotting.py]:
```
if 'decision_type' == 'no_greater':
  left_child <= 'threshold'
  right_child > 'threshold'
else if 'decision_type' == 'is':
  left_child is 'threshold'
  right_child isn't 'threshold'
```

Comment (wxchan) [inline python-package/lightgbm/plotting.py]: Fixed.

**Comment by wxchan** [comment]
Graphviz are not executables on Travis, so there is no way to test it. I just put an empty function with a skip message there to notice others.

**Comment by wxchan** [comment]
Here is an example how the plot tree looks like in `examples/python-guide/plot_example.py`.
plot_example

File: python-package/lightgbm/plotting.py (hunk 1 above)
Comment (wxchan) [inline python-package/lightgbm/plotting.py]: Wait a minute..there is a bug.

**Comment by henry0312** [comment]
Wow, this looks good!
Is it easy to show a feature name instead of the index?

**Comment by wxchan** [comment]
@guolinke finished.

@henry0312 feature name is saved in Dataset, not in Booster now. We can investigate how to use feature name instead of index after this pr.

**Comment by henry0312** [comment]
> feature name is saved in Dataset, not in Booster now.

I see...🤔

**Comment by guolinke** [comment]
@wxchan I think we can add feature names to booster as well. 
Actually inside boosting class has feature_names

**Comment by wxchan** [comment]
I use feature_name in dump_model for now. @guolinke @henry0312 
plot_example

File: python-package/lightgbm/plotting.py [hunk 2]
```diff
@@ -110,3 +121,118 @@ def plot_importance(booster, ax=None, height=0.2,
...
+        raise IndexError('tree_index is out of range.')
+
+    graph = Digraph(graph_attr=graph_attr, node_attr=node_attr, edge_attr=edge_attr)
+
+    if show_info is None:
+        show_info = []
+    ret = _to_graphviz(graph, tree_info, show_info, feature_names)
+
+    s = BytesIO()
+    s.write(ret.pipe(format='png'))
```
Comment (henry0312) [inline python-package/lightgbm/plotting.py]: only support png?

Comment (henry0312) [inline python-package/lightgbm/plotting.py]: oh, sorry.
I misunderstood. Please ignore.

**Comment by henry0312** [comment]
seems good :)

**Comment by henry0312** [comment]
@chivee I think it's better to use squash and merge commit to merge PR like @guolinke.

**Comment by guolinke** [comment]
It seems the merge of this PR add many additional commit log in master branch.
I will force update the master branch now. 
@wxchan  can you re-create this PR again ?

**Comment by chivee** [comment]
@henry0312, I was assuming GitHub using squash and merge as a default option.

**Comment by chivee** [comment]
@guolinke , please using reset soft rather than 'hard  mode' as possible

**Pull request merged**